### ISSUE TITLE
Tile Rules do not take into account other tiles in same tilemap

### ISSUE THREAD

**Issue by nicbiondi** [Unity-Technologies/2d-extras#27]
Rule tiles do not take into account positions of other tiles in the same grid even if they are on the same tilemap. It would be nice to have a toggle to allow rules to take other tiles (non rule tile or normal tiles) into account when deciding which sprite to use.

**Comment by edwardrowe** [comment]
On our project (Sparklite) we modified the RuleTile to have what we call "sibling" tiles. When deciding if a rule matches, instead of just comparing to 'this' we also see if it's a sibling.

**Comment by edwardrowe** [comment]
We also have siblings PER RULE but I'm wondering if that's a hack cause it feels hacky.

**Comment by nicbiondi** [comment]
Dang nice work edward!
 let's hope this project gets some more love. It really is a beauty. I would also accept giving the Generic Rule tile more than 9 siblings. I know this would be overkill for some uses but for mine it would be nice to be able to set rules for two away rather than just one distance.

**Comment by nicbiondi** [comment]
Hi ChuanXin, I agree with this. the default version should be simple to grasp, and it is! This is the first time I have used the RuleTile and ran into the issue of needing to extend the functionality so I thought I would mention it as I believe it will be necessary for many users.

**Comment by johnsoncodehk** [comment]
I agree that we should keep the current RuleTile simple to use, so I made a version that is extensible and backwards compatible, allowing you to customize fields and rules.

Commit: https://github.com/johnsoncodehk/2d-extras/commit/6e879e7fe05649619911aae8a52cd6dc46877aab

Extension code:
```csharp
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.Tilemaps;

[CreateAssetMenu]
public class PFTile : RuleTile {

	public bool isGround;

	protected override List<int> GetNeighborRules() {
		List<int> neighbors = base.GetNeighborRules();
		neighbors.Add(3); // Null
		neighbors.Add(4); // NotNullAndThis
		neighbors.Add(5); // IsGround
		return neighbors;
	}
	public override bool RuleMatche(int neighbor, TileBase neighboringTile) {
		switch (neighbor) {
			case 3: return neighboringTile == null;
			case 4: return neighboringTile != null && neighboringTile != this;
			case 5: return (neighboringTile as PFTile).isGround;
		}
		return base.RuleMatche(neighbor, neighboringTile);
	}
}
```
```csharp
using UnityEngine;
using UnityEditor;

[CustomEditor(typeof(PFTile))]
[CanEditMultipleObjects]
public class PFTileEditor : RuleTileEditor {
	public override void OnInspectorGUI() {
		EditorGUILayout.PropertyField(serializedObject.FindProperty("isGround"));
		base.OnInspectorGUI();
	}
	public override void RuleOnGUI(Rect rect, Vector2Int pos, int neighbor) {
		switch (neighbor) {
			case 3: GUI.Label(rect, "N"); break;
			case 4: GUI.Label(rect, "X"); break;
			default: base.RuleOnGUI(rect, pos, neighbor); break;
		}
	}
}
```

There may be an easier way, but I hope my direction is correct.

**Comment by johnsoncodehk** [comment]
I have submitted a pull request for this requirement: https://github.com/Unity-Technologies/2d-extras/pull/37

**Comment by nicbiondi** [comment]
ooh nice! is this pulled? I can't seem to tell. thanks for this!

**Comment by johnsoncodehk** [comment]
This branch can be found here:
https://github.com/johnsoncodehk/2d-extras/tree/rule-tile-extend

When the author did not confirm the merger of these changes, I suggest not to update it for the time being because everything may change.
